### PR TITLE
Adding beta of TimeDistributed

### PR DESCRIPTION
**Pull request by lukedeo**
Just opening this to start a convo -- @fchollet if this is something you'd be interested in merging, let me know. I can write some tests + docs in that case. Similar to some changes + comments in #1456.

### CODE REVIEW

**Comment by fchollet** [comment]
Schematically this is what we had in mind. There are a few technical issues though.

> I can write some tests + docs

Just a docstring. Tests will have to cover the most common use cases: application to Dense and to Convolution2D.

File: keras/layers/core.py [hunk 1]
```diff
@@ -995,6 +995,86 @@ def get_config(self):
...
+
+    @property
+    def output_shape(self):
+        input_shape = self.input_shape
+        return (input_shape[0], input_shape[1]) + self.layer.output_shape[1:]
+
+    def get_output(self, train=False):
+        def format_shape(shape):
+            if K._BACKEND == 'tensorflow':
+                return map(int, shape)
```
Comment (fchollet) [inline keras/layers/core.py]: I don't think it's acceptable to have backend-specific statements if it can be avoided. Also this would fail most of the time because usually the input shape contains `None`.

File: keras/layers/core.py [hunk 2]
```diff
@@ -995,6 +995,86 @@ def get_config(self):
...
+        self.input_ndim = len(layer.output_shape)
+        super(TimeDistributed, self).set_previous(layer)
+
+    def build(self):
+        try:
+            self.input_ndim = len(self.previous.input_shape)
+        except AttributeError:
+            self.input_ndim = len(self.input_shape)
+
+        self.layer.set_input_shape((None, ) + self.input_shape[2:])
```
Comment (fchollet) [inline keras/layers/core.py]: The batch size dimension might not be `None`. It has to be specified in quite a few situations (e.g. usage of the present layer in a stateful network).

File: keras/layers/core.py (hunk 1 above)
Comment (lukedeo) [inline keras/layers/core.py]: Right, fair point. Any ideas as to how to make the batch + timestep dimension flattening work between both backends? At a bit of a loss...

Comment (fchollet) [inline keras/layers/core.py]: We could add a backend method to do it, with custom implementations in each backend...

**Comment by fchollet** [comment]
Won't merge, a more general version of the layer is in preparation.

**Comment by lukedeo** [comment]
:+1: awesome
